### PR TITLE
Added FP16 reference implementation

### PR DESCRIPTION
**Pull request by l-bat**

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
@l-bat could you please clarify what does it fix? Why branch is called `unstable`?

**Comment by l-bat** [comment]
> @l-bat could you please clarify what does it fix? Why branch is called `unstable`?

@ilya-lavrenov this is old name of branch. When I created it I thought that some tests failed on Mac due to unstable calculation.
But these changes relate to switch to `ngraph::float16` instead of `half_float` and adding  FP16 reference implementation for some ops.

File: modules/arm_plugin/tests/functional/shared_tests_instances/skip_tests_config.cpp [hunk 1]
```diff
@@ -16,6 +16,7 @@ std::vector<std::string> disabledTestPatterns() {
         R"(.*Eltwise.*eltwiseOpType=Mod.*netPRC=FP16.*)", // Failed
         ".*PreprocessTest.*", // Does not cover all needed cases
         ".*GRUCellTest.*decomposition0.*",  // GruCell should be decomposed
-        R"(.*ConstantResultSubgraphTest.*inPrc=(I8|U64|I64|BOOL).*)" // Unsupported precisions
+        R"(.*ConstantResultSubgraphTest.*inPrc=(I8|U64|I64|BOOL).*)", // Unsupported precisions
+        ".*ActivationLayerTest.*CompareWithRefs/Tan_.*netPRC=FP16.*" // Failed (a small input change leads to a large output change)
```
Comment (ilya-lavrenov) [inline modules/arm_plugin/tests/functional/shared_tests_instances/skip_tests_config.cpp:20]: put under `#ifdef __ARM_FEATURE_FP16_VECTOR_ARITHMETIC`?

File: modules/arm_plugin/src/arm_converter/arm_converter_region_yolo.cpp [hunk 2]
```diff
@@ -19,9 +19,10 @@ template<> Converter::Conversion::Ptr Converter::Convert(const opset::RegionYolo
                                     node.get_mask());
     };
 
-    if (node.get_input_element_type(0) != ngraph::element::f32) {
-        THROW_IE_EXCEPTION << "Unsupported Type: " << node.get_input_element_type(0);
+    switch (node.get_input_element_type(0)) {
+        case ngraph::element::Type_t::f16 : return make(ngraph::runtime::reference::region_yolo<ngraph::float16>);
```
Comment (ilya-lavrenov) [inline modules/arm_plugin/src/arm_converter/arm_converter_region_yolo.cpp:23]: @apankratovantonp should not we call these references under `#ifdef __ARM_FEATURE_FP16_VECTOR_ARITHMETIC`? They seem to be not needed for FP32 execution

Comment (ilya-lavrenov) [reply modules/arm_plugin/src/arm_converter/arm_converter_region_yolo.cpp:23]: And maybe in opposite case - FP32 kernels are not needed for FP16 case..

**Comment by ilya-lavrenov** [comment]
I also see the following failures:
```
[  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/LeakyRelu_IS=(1.50)_AS=(1)_ConstantsValue=(0.01)_netPRC=FP16_inPRC=UNSPECIFIED_out
PRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((4, { 0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 50 }, { 1 }), "CPU")            
[  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/LeakyRelu_IS=(1.50)_AS=(50)_ConstantsValue=(0.01)_netPRC=FP16_inPRC=UNSPECIFIED_ou
tPRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((4, { 0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 50 }, { 50 }), "CPU")          
[  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/LeakyRelu_IS=(1.128)_AS=(1)_ConstantsValue=(0.01)_netPRC=FP16_inPRC=UNSPECIFIED_ou
tPRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((4, { 0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 128 }, { 1 }), "CPU")          
[  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/LeakyRelu_IS=(1.128)_AS=(128)_ConstantsValue=(0.01)_netPRC=FP16_inPRC=UNSPECIFIED_
outPRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((4, { 0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 128 }, { 128 }), "CPU")      
[  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/PReLu_IS=(1.50)_AS=(1)_ConstantsValue=(-0.01)_netPRC=FP16_inPRC=UNSPECIFIED_outPRC
=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((27, { -0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 50 }, { 1 }), "CPU")             
[  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/PReLu_IS=(1.50)_AS=(50)_ConstantsValue=(-0.01)_netPRC=FP16_inPRC=UNSPECIFIED_outPR
C=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((27, { -0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 50 }, { 50 }), "CPU")           
[  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/PReLu_IS=(1.128)_AS=(1)_ConstantsValue=(-0.01)_netPRC=FP16_inPRC=UNSPECIFIED_outPR
C=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((27, { -0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 128 }, { 1 }), "CPU")           
[  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/PReLu_IS=(1.128)_AS=(128)_ConstantsValue=(-0.01)_netPRC=FP16_inPRC=UNSPECIFIED_out
PRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((27, { -0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 128 }, { 128 }), "CPU")       
[  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(5.1.10)_secondaryInputType=CONSTANT_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=A
NY_outL=ANY_trgDev=CPU, where GetParam() = ({ 5, 1, 10 }, CONSTANT, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                            
[  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(5.1.10)_secondaryInputType=PARAMETER_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=
ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 5, 1, 10 }, PARAMETER, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                          
[  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(1.1.10)_secondaryInputType=CONSTANT_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=A
NY_outL=ANY_trgDev=CPU, where GetParam() = ({ 1, 1, 10 }, CONSTANT, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                            
[  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(1.1.10)_secondaryInputType=PARAMETER_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=
ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 1, 1, 10 }, PARAMETER, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                          
[  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(20.1.10)_secondaryInputType=CONSTANT_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=
ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 20, 1, 10 }, CONSTANT, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                          
[  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(20.1.10)_secondaryInputType=PARAMETER_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL
=ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 20, 1, 10 }, PARAMETER, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                        
[  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(20.20.10)_secondaryInputType=CONSTANT_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL
=ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 20, 20, 10 }, CONSTANT, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                        
[  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(20.20.10)_secondaryInputType=PARAMETER_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_in
L=ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 20, 20, 10 }, PARAMETER, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                      
[  FAILED  ] smoke_Reduce/ReduceOpsLayerTest.CompareWithRefs/IS=(5.6.10.11)_axes=(2.3)_opType=VECTOR_type=Prod_KeepDims_netPRC=FP32_inPRC=UNSPECIFIED_outPRC=UNS
PECIFIED_inL=ANY_trgDev=CPU, where GetParam() = ({ 2, 3 }, VECTOR, true, Prod, FP32, UNSPECIFIED, UNSPECIFIED, ANY, { 5, 6, 10, 11 }, "CPU")                    
[  FAILED  ] smoke_Reduce/ReduceOpsLayerTest.CompareWithRefs/IS=(5.6.10.11)_axes=(2.3)_opType=VECTOR_type=Prod_netPRC=FP32_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_
inL=ANY_trgDev=CPU, where GetParam() = ({ 2, 3 }, VECTOR, false, Prod, FP32, UNSPECIFIED, UNSPECIFIED, ANY, { 5, 6, 10, 11 }, "CPU")                            
                                                                                                                                                                
18 FAILED TESTS                                                                                                                                                 
  YOU HAVE 7 DISABLED TESTS                                                                                                                                     
```

**Comment by l-bat** [comment]
> I also see the following failures:
> 
> ```
> [  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/LeakyRelu_IS=(1.50)_AS=(1)_ConstantsValue=(0.01)_netPRC=FP16_inPRC=UNSPECIFIED_out
> PRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((4, { 0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 50 }, { 1 }), "CPU")            
> [  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/LeakyRelu_IS=(1.50)_AS=(50)_ConstantsValue=(0.01)_netPRC=FP16_inPRC=UNSPECIFIED_ou
> tPRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((4, { 0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 50 }, { 50 }), "CPU")          
> [  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/LeakyRelu_IS=(1.128)_AS=(1)_ConstantsValue=(0.01)_netPRC=FP16_inPRC=UNSPECIFIED_ou
> tPRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((4, { 0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 128 }, { 1 }), "CPU")          
> [  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/LeakyRelu_IS=(1.128)_AS=(128)_ConstantsValue=(0.01)_netPRC=FP16_inPRC=UNSPECIFIED_
> outPRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((4, { 0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 128 }, { 128 }), "CPU")      
> [  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/PReLu_IS=(1.50)_AS=(1)_ConstantsValue=(-0.01)_netPRC=FP16_inPRC=UNSPECIFIED_outPRC
> =UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((27, { -0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 50 }, { 1 }), "CPU")             
> [  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/PReLu_IS=(1.50)_AS=(50)_ConstantsValue=(-0.01)_netPRC=FP16_inPRC=UNSPECIFIED_outPR
> C=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((27, { -0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 50 }, { 50 }), "CPU")           
> [  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/PReLu_IS=(1.128)_AS=(1)_ConstantsValue=(-0.01)_netPRC=FP16_inPRC=UNSPECIFIED_outPR
> C=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((27, { -0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 128 }, { 1 }), "CPU")           
> [  FAILED  ] smoke_Activation_Basic_Prelu/ActivationLayerTest.CompareWithRefs/PReLu_IS=(1.128)_AS=(128)_ConstantsValue=(-0.01)_netPRC=FP16_inPRC=UNSPECIFIED_out
> PRC=UNSPECIFIED_inL=ANY_outL=ANY_trgDev=CPU, where GetParam() = ((27, { -0.01 }), FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, ({ 1, 128 }, { 128 }), "CPU")       
> [  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(5.1.10)_secondaryInputType=CONSTANT_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=A
> NY_outL=ANY_trgDev=CPU, where GetParam() = ({ 5, 1, 10 }, CONSTANT, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                            
> [  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(5.1.10)_secondaryInputType=PARAMETER_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=
> ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 5, 1, 10 }, PARAMETER, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                          
> [  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(1.1.10)_secondaryInputType=CONSTANT_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=A
> NY_outL=ANY_trgDev=CPU, where GetParam() = ({ 1, 1, 10 }, CONSTANT, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                            
> [  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(1.1.10)_secondaryInputType=PARAMETER_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=
> ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 1, 1, 10 }, PARAMETER, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                          
> [  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(20.1.10)_secondaryInputType=CONSTANT_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL=
> ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 20, 1, 10 }, CONSTANT, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                          
> [  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(20.1.10)_secondaryInputType=PARAMETER_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL
> =ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 20, 1, 10 }, PARAMETER, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                        
> [  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(20.20.10)_secondaryInputType=CONSTANT_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_inL
> =ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 20, 20, 10 }, CONSTANT, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                        
> [  FAILED  ] smoke_GatherTree/GatherTreeLayerTest.CompareWithRefs/IS=(20.20.10)_secondaryInputType=PARAMETER_netPRC=FP16_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_in
> L=ANY_outL=ANY_trgDev=CPU, where GetParam() = ({ 20, 20, 10 }, PARAMETER, FP16, UNSPECIFIED, UNSPECIFIED, ANY, ANY, "CPU")                                      
> [  FAILED  ] smoke_Reduce/ReduceOpsLayerTest.CompareWithRefs/IS=(5.6.10.11)_axes=(2.3)_opType=VECTOR_type=Prod_KeepDims_netPRC=FP32_inPRC=UNSPECIFIED_outPRC=UNS
> PECIFIED_inL=ANY_trgDev=CPU, where GetParam() = ({ 2, 3 }, VECTOR, true, Prod, FP32, UNSPECIFIED, UNSPECIFIED, ANY, { 5, 6, 10, 11 }, "CPU")                    
> [  FAILED  ] smoke_Reduce/ReduceOpsLayerTest.CompareWithRefs/IS=(5.6.10.11)_axes=(2.3)_opType=VECTOR_type=Prod_netPRC=FP32_inPRC=UNSPECIFIED_outPRC=UNSPECIFIED_
> inL=ANY_trgDev=CPU, where GetParam() = ({ 2, 3 }, VECTOR, false, Prod, FP32, UNSPECIFIED, UNSPECIFIED, ANY, { 5, 6, 10, 11 }, "CPU")                            
>                                                                                                                                                                 
> 18 FAILED TESTS                                                                                                                                                 
>   YOU HAVE 7 DISABLED TESTS                                                                                                                                     
> ```

Yes, these tests will be fixed in another PR.

Fix for GatherTree: openvinotoolkit/openvino#4917

File: modules/arm_plugin/src/arm_converter/arm_converter_region_yolo.cpp (hunk 2 above)
Comment (apankratovantonp) [reply modules/arm_plugin/src/arm_converter/arm_converter_region_yolo.cpp:23]: I preferer to leave binary size optimization to selective build feature and enable it for all types.

**Review by apankratovantonp — APPROVED** [review]

File: modules/arm_plugin/src/arm_converter/arm_converter_region_yolo.cpp (hunk 2 above)
Comment (ilya-lavrenov) [reply modules/arm_plugin/src/arm_converter/arm_converter_region_yolo.cpp:23]: You will have dead code. E.g. FP16 kernels physically cannot be run on FP32 arch.

**Comment by ilya-lavrenov** [comment]
@l-bat please, resolve merge conflicts.

**Pull request merged**